### DISCUSSION TITLE
About customized middleware

### DISCUSSION THREAD

**Discussion by hulinux** [directus/directus#9930] [Archive]
How to add customized middleware, such as encryption and decryption of http requests

**Comment by rijkvanzanten** [comment]
You can use the init hooks to tap into the various stages of the API creation:

https://docs.directus.io/extensions/hooks/#init

For your particular use case, you're probably looking to use middlewares.before or middlewares.after 👍🏻

**Reply by abdonrd** [reply]
@rijkvanzanten with `middlewares.before` can I override a core middleware?

**Reply by rijkvanzanten** [reply]
No, that won't allow you to override Directus' core middleware. That being said, IIRC express stores all the registered middlewares in an array, so you might be able to use `middlewares.after` to modify the stack of registered middlewares afterwards 🤔  

Out of personal curiosity, what middleware are you trying to override, and for what use case? 🙂

**Reply by abdonrd** [reply]
The `api/src/middleware/authenticate.ts`.

To authenticate some API users against an external API ([instead of the one that Directus brings](https://github.com/directus/directus/blob/27e863f1973994b323b5a5d037e89a477191bcbf/api/src/middleware/authenticate.ts#L25-L68)), and assign a default role to those users. I guess it's a rather peculiar case.

**Reply by rijkvanzanten** [reply]
Would that be better served with a custom auth adapter, or is this too specific of a use case? That way, you could disable the default login, show a proper login button / form for whatever external API you use, and then avoid having to override internal logic 🤔 

(right now, we have oauth2 / openid / ldap integrated, but we're looking into making those an extension type as well, which could be a different solution for the same problem)

**Reply by abdonrd** [reply]
But we want to keep all the logic of Directus, since we also use the App, etc.

Our use case is "just" add an `} if else (isCustomJWT(req.token)) { ... }` here:

https://github.com/directus/directus/blob/27e863f1973994b323b5a5d037e89a477191bcbf/api/src/middleware/authenticate.ts#L46

And we don't store those users in Directus.
I guess it's a bit of a weird implementation, I've never done it before.

**Reply by rijkvanzanten** [reply]
> And we don't store those users in Directus.

Ahh that's the ticket, gotcha. I was thinking you were implementing a custom way to login to a Directus user 👍🏻

**Reply by rijkvanzanten** [reply]
We're actually currently implementing a hook that would allow you to parse the incoming token yourself first, and then tell Directus what accountability token to use from that point onwards, that might actually be exactly what you're looking for here!

**Reply by abdonrd** [reply]
Sounds good, yes!

**Reply by abdonrd** [reply]
@rijkvanzanten thanks to your PR https://github.com/directus/directus/pull/11942 we have been able to remove the patch (with patch-package) we had. Thanks! 👏🏻 

```javascript
filter('authenticate', async (accountability, { req }) => {
  const customUser = await getUserFromExternalAccessToken(req.token);

  if (customUser) {
    return {
      ...accountability,
      custom_user: customUser.id,
      role: process.env.CUSTOM_ROLE_ID,
    };
  }

  return accountability;
});
```

Another related question...
There is no option to add a custom dynamic variable, right?
Because we have one with another patch (it's the last one we have)

https://github.com/directus/directus/blob/079cf6042569fb67b772e6708a07aa3b9902e102/packages/shared/src/utils/is-dynamic-variable.ts#L1

https://github.com/directus/directus/blob/079cf6042569fb67b772e6708a07aa3b9902e102/packages/shared/src/utils/parse-filter.ts#L83-L87

Here we add a:

```javascript
if (value.startsWith('$CURRENT_CUSTOM_USER')) { 
  if (value === '$CURRENT_CUSTOM_USER') return accountability?.custom_user ?? null; 
  return get(context, value, null); 
} 
```

**Reply by rijkvanzanten** [reply]
Good one! There's currently no way to add any custom dynamic variables. I'm curious how we might make that work 🤔 It could be a custom hook potentially to process unknown dynamic variables, or maybe another way to register "i want var X to mean Y"

**Reply by abdonrd** [reply]
Yeah, I also don't know the best way, but should be really nice to have the option to do it!

**Reply by abdonrd** [reply]
Hi again @rijkvanzanten!

Regarding to this: https://github.com/directus/directus/discussions/9930#discussioncomment-2322535

There is any option to use the `isDirectusJWT` inside the new `authenticate` filter hook?

**Reply by abdonrd** [reply]
We want to wrapper it with the `isDirectusJWT` like:

```javascript
filter('authenticate', async (accountability, { req }) => {
  if (req.token && !isDirectusJWT(req.token)) {
    const customUser = await getUserFromExternalAccessToken(req.token);
  
    if (customUser) {
      return {
        ...accountability,
        custom_user: customUser.id,
        role: process.env.CUSTOM_ROLE_ID,
      };
    }
  }

  return accountability;
});
```

**Reply by abdonrd** [reply]
Oh, yes! We can here:

```javascript
const isDirectusJWT = require('directus/dist/utils/is-directus-jwt')
```

**Reply by abdonrd** [reply]
Oops... It doesn't work:

```javascript
const isDirectusJWT = require('directus/dist/utils/is-directus-jwt');
```

```
18:32:47 ⚠️  Couldn't register hook "extension-name"
18:32:47 ⚠️  Cannot find module '/Users/abdonrd/Work/IBM/.../node_modules/directus/dist/dist/utils/is-directus-jwt.js'
18:32:47 ✨ Loaded extensions: extension-name
18:32:47 ⚠️  PUBLIC_URL should be a full URL
```

**Reply by abdonrd** [reply]
Oh! You use the `exports` field in the `package.json`!
It works if I import like this:

```javascript
const isDirectusJWT = require('directus/utils/is-directus-jwt').default;
```

Note the non `dist/` in the path and the `.default`.

**Reply by jeffersonsolorzano** [reply]
Hi @abdonrd, I' trying to use that func inside a bundle, but us seems impossible to use, do i have to use any configuration inside my package.json?

**Comment by piotr-cz** [comment]
Here is link to comment how to add custom middleware: https://github.com/directus/directus/discussions/16191#discussioncomment-3978662

This is my implementation in TypeScript after `npm init directus-extension`:

```ts
// src/index.ts
import { defineHook } from '@directus/extensions-sdk';
import type { Application } from 'express';

export default defineHook(({ init }, { env, logger }) => {
	// Note: For ping the middlewares.after event is not triggered
	init('middlewares.after', ({ app }: Record<'app', Application>) => {
		app.use((_req, res, next) => {
			// Modify response
			// ...
			next();
		});
	});
});

```
